octobot_trading/personal_data/portfolios/portfolio_util.py

```diff
@@ -69,4 +70,103 @@ def get_draw_down(exchange_manager):
             current_draw_down = 100 - (portfolio_history[-1] / (max(portfolio_history) / 100))
 
             draw_down = current_draw_down if current_draw_down > draw_down else draw_down
-    return draw_down
+    return round(draw_down, 3)
+
+
+async def get_coefficient_of_determination_data(transactions, start_balance,
+                                                use_high_instead_of_end_balance=True,
+                                                x_as_trade_count=True):
+    if transactions:
+        # get realized portfolio history
+
+        pnl_history = [start_balance]
+        pnl_history_times = []
+        for transaction in transactions:
+            current_pnl = None
+            if hasattr(transaction, "quantity"):
+                current_pnl = float(transaction.quantity)
+                pnl_history_times.append(transaction.creation_time)
+            elif hasattr(transaction, 'realised_pnl'):
+                current_pnl = float(transaction.realised_pnl)
+                pnl_history_times.append(transaction.creation_time)
+            elif isinstance(transaction, dict):
+                if transaction["quantity"]:
+                    current_pnl = transaction["quantity"]
+                    pnl_history_times.append(transaction["x"])
+                elif transaction['realised_pnl']:
+                    current_pnl = transaction['realised_pnl']
+                    pnl_history_times.append(transaction["x"])
+
+            if current_pnl is not None:
+                pnl_history.append(pnl_history[-1] + current_pnl)
+
+        if pnl_history_times:
+
+            pnl_history_for_every_candle = None
+            # either use trade to trade basis or candle to candle basis
+            if x_as_trade_count is True:
+                start_time = pnl_history_times[0]
+                pnl_history_times.insert(0, start_time)
+                end_time = pnl_history_times[-1]
+                data_length = len(pnl_history)
+
+            else:  # calculate pnl history for every candle
+                raise NotImplementedError("x_as_trade_count=False is not implemented")
+            # calculate best case data (exponential growth)
+            end_balance = pnl_history[-1]
+
+            if use_high_instead_of_end_balance:
+                end_value = max(pnl_history)
+            else:  # use the end balance
+                end_value = end_balance
+
+            def get_ideal_value(linear_growth, adj1, adj2):
+                return ((linear_growth + adj1) ** pw) * adj2
+
+            x = [start_time, end_time]
+            y = [start_balance, end_value]
+
+            pw = 15
+            A = numpy.exp(numpy.log(y[0] / y[1]) / pw)
```

Comment:
😕

Reply:
im not fully sure what those lines do. I got it from here: https://stackoverflow.com/a/33190889

Reply:
@techfreaque did you test lines? How can we ensure it is working?